02_Util/src/security/SignedMeterValuesUtil.ts

```diff
@@ -0,0 +1,209 @@
+import {
+  IChargingStationSecurityInfoRepository,
+  IDeviceModelRepository,
+  sequelize,
+} from '@citrineos/data';
+import {
+  IFileAccess,
+  MeterValueType,
+  SignedMeterValuesConfig,
+  SignedMeterValueType,
+  SystemConfig,
+} from '@citrineos/base';
+import { ILogObj, Logger } from 'tslog';
+import * as crypto from 'node:crypto';
+
+/**
+ * Util to process and validate signed meter values.
+ */
+export class SignedMeterValuesUtil {
+  private readonly _fileAccess: IFileAccess;
+  private readonly _logger: Logger<ILogObj>;
+  private readonly _deviceModelRepository: IDeviceModelRepository;
+  private readonly _chargingStationSecurityInfoRepository: IChargingStationSecurityInfoRepository;
+
+  private readonly _signedMeterValuesConfiguration:
+    | SignedMeterValuesConfig
+    | undefined;
+
+  /**
+   * @param {IFileAccess} [fileAccess] - The `fileAccess` allows access to the configured file storage.
+   *
+   * @param {SystemConfig} config - The `config` contains the current system configuration settings.
+   *
+   * @param {Logger<ILogObj>} [logger] - The `logger` represents an instance of {@link Logger<ILogObj>}.
+   *
+   * @param {IDeviceModelRepository} [deviceModelRepository] - The `deviceModelRepository` represents a repository for accessing and manipulating variable attribute data.
+   *
+   */
+  constructor(
+    fileAccess: IFileAccess,
+    config: SystemConfig,
+    logger: Logger<ILogObj>,
+    deviceModelRepository: IDeviceModelRepository,
+  ) {
+    this._fileAccess = fileAccess;
+    this._logger = logger;
+    this._deviceModelRepository = deviceModelRepository;
+    this._chargingStationSecurityInfoRepository =
+      new sequelize.SequelizeChargingStationSecurityInfoRepository(
+        config,
+        logger,
+      );
+
+    this._signedMeterValuesConfiguration =
+      config.modules.transactions.signedMeterValuesConfiguration;
+  }
+
+  public async validateMeterValues(
+    stationId: string,
+    meterValues: [MeterValueType, ...MeterValueType[]],
+  ): Promise<boolean> {
+    let validMeterValues = true;
+
+    const expectPublicKey =
+      await this._deviceModelRepository.readAllByQuerystring({
+        stationId,
+        component_name: 'OCPPCommCtrlr',
+        variable_name: 'PublicKeyWithSignedMeterValue',
+      });
+
+    for (const meterValue of meterValues) {
+      if (expectPublicKey.length > 0 && expectPublicKey[0].value !== 'Never') {
+        for (const sampledValue of meterValue.sampledValue) {
+          const signedMeterValue = sampledValue.signedMeterValue;
+
+          if (!signedMeterValue) {
+            continue;
+          }
+
+          if (
+            signedMeterValue.publicKey &&
+            signedMeterValue.publicKey.length > 0
+          ) {
+            const incomingPublicKeyIsValid =
+              await this.isMeterValueSignatureValid(signedMeterValue);
+
+            if (
+              this._signedMeterValuesConfiguration &&
```

Comment:
Why would this CSMS config need to be present in order for the signed meter values to be deemed valid?

At this point, we already validated that the signed meter value sent by the charger is valid so why does it matter if we have this config present?

Also, if this config is truly required for any sort of signed meter values validation, maybe we should have it be one of the first checks we perform in the `validateMeterValues` method?

Reply:
we validate mostly as a sanity check to make sure some really messed up data isn't just being processed (although, we save it, so i suppose it'll still end up somewhere in our system.)

checking if the config is present is only necessary to help us create that database entry that stores the link between the charging station + the public key (file id). right now, the database entry stores the public key file id, rather than the actual public key itself, and the only way to derive the public key file id is by checking the config 😅